OSDOCS#8867: Adding failure domains to Nutanix docs (#69618)
openshift/openshift-docs: modules/post-installation-adding-nutanix-failure-domains-compute-machines.adoc

// Module included in the following assemblies:
//
// * post_installation_configuration/adding-nutanix-failure-domains.adoc

:_mod-docs-content-type: PROCEDURE
[id="post-installation-adding-nutanix-failure-domains-compute-machines_{context}"]
= Distributing compute machines across failure domains

> You can distribute compute machines across Nutanix failure domains by:

Reviewer: Per ISG
```suggestion
You can distribute compute machines across Nutanix failure domains by performing either of the following tasks:
```

Author: Done. Thank you.

* Modifying existing compute machine sets.
* Creating new compute machine sets.

> The following procedure details how to distribute compute machines across failure domains by modifying existing control plane machines sets. For more information on creating a control plane machine set, see "Additional resources".

Reviewer 2: Noticed while getting started on my new module:
```suggestion
The following procedure details how to distribute compute machines across failure domains by modifying existing compute machine sets. For more information on creating a compute machine set, see "Additional resources".
```

Author: Done. Thank you.

.Prerequisites

> * You have configured the failure domains in the cluster's Infrastructure custom resource (CR).

Reviewer: ```suggestion
* You have configured the failure domains in the cluster's Infrastructure custom resource definition (CRD).
```

Author: This was raised early as part of an SME review. The decision was to go with CR.

https://github.com/openshift/openshift-docs/pull/69618#discussion_r1457598396

> * You have configured the failure domains in the cluster's Infrastructure custom resource definition (CR).

Reviewer 3: Remove "definition". Should be "in the cluster's Infrastructure custom resource (CR)".

Author: > Remove "definition". Should be "in the cluster's Infrastructure custom resource (CR)".

I am going to check with @jeana-redhat when she returns from PTO next week. I have seen them refer to this as "infrastructure custom resource definition" elsewhere [1] . Additionally - existing published docs [2] (see step 1) refer to `infrastructures.config.openshift.io` as "custom resource definition" (CRD). I will close on this terminology and update as required.

[1] https://github.com/openshift/openshift-docs/pull/70056/files#diff-1f902b9f28f2684b458efb2ab87443a9c36f86c8b5caffea4783dc122bb52183R13
[2] https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/post_installation_configuration/post-install-vsphere-zones-regions-configuration#specifying-regions-zones-infrastructure-vsphere_post-install-vsphere-zones-regions-configuration

Reviewer 3: @JoelSpeed Can yo clarify here: should "infrastructure custom resource definition (CRD)" or "infrastructure custom resource (CR)" be used here? I think it should be CR without "definition", since CR refers to the custom resource object.

Reviewer 4: It should be `cutom resource (CR)` to refer to the single instance of the custom resource

Reviewer 2: I copied the use in my PR from this PR :) I will update mine accordingly to remove "definition"

Author: Also done.

.Procedure

> . Run the following command to view the cluster's Infrastructure CR.

Reviewer: ```suggestion
. Run the following command to view the cluster's Infrastructure CRD.
```

Author: Please see https://github.com/openshift/openshift-docs/pull/69618#discussion_r1471804179

Author: Please see https://github.com/openshift/openshift-docs/pull/69618#discussion_r1471804179

+
[source,terminal]
----
$ oc describe infrastructures.config.openshift.io cluster
----
. For each failure domain (`platformSpec.nutanix.failureDomains`), note the cluster's UUID, name, and subnet object UUID. These values are required to add a failure domain to a compute machine set.
. List the compute machine sets in your cluster by running the following command:
+
[source,terminal]
----
$ oc get machinesets -n openshift-machine-api
----

> . Edit the first compute machine set by running the following command:

Reviewer: Not sure _first_ is clear here. The first machine set you want to edit? The first machine set in the command output? Maybe it doesn't matter.....
```suggestion
. Edit a compute machine set by running the following command:
```

+
[source,terminal]
----
$ oc edit machineset <machineset_name> -n openshift-machine-api
----

> . Configure the compute machine set to use the the first failure domain by adding the following to the `spec.template.spec.providerSpec.value` stanza:

Reviewer: Same here, maybe?
```suggestion
. Configure the compute machine set to use a first failure domain by adding the following to the `spec.template.spec.providerSpec.value` stanza:
```

Author: This was raised early as part of an SME review. The decision was to go with CR.

https://github.com/openshift/openshift-docs/pull/69618#discussion_r1457598396

Author: I removed the extra `the`.

+
[NOTE]
====

> Be sure that the values you specify for the `cluster` and `subnets` fields match the values that were configured in the `failureDomains` stanza in the cluster's Infrastructure CR.

Reviewer: ```suggestion
Be sure that the values you specify for the `cluster` and `subnets` fields match the values that were configured in the `failureDomains` stanza in the cluster's Infrastructure CRD.
```

Author: This was raised early as part of an SME review. The decision was to go with CR.

https://github.com/openshift/openshift-docs/pull/69618#discussion_r1457598396

          failureDomain:
            name: <failure_domain_name_1>
          cluster:
            type: uuid
            uuid: <prism_element_uuid_1>
          subnets:
          - type: uuid
            uuid: <prism_element_network_uuid_1>
# ...
----
. Note the value of `spec.replicas`, as you need it when scaling the machine set to apply the changes.
. Save your changes.
. List the machines that are managed by the updated compute machine set by running the following command:
+
[source,terminal]
----
$ oc get -n openshift-machine-api machines -l machine.openshift.io/cluster-api-machineset=<machine_set_name>
----
. For each machine that is managed by the updated compute machine set, set the `delete` annotation by running the following command:
+
[source,terminal]
----
$ oc annotate machine/<machine_name_original_1> \
  -n openshift-machine-api \
  machine.openshift.io/delete-machine="true"
----
. Scale the compute machine set to twice the number of replicas by running the following command:
+
[source,terminal]
----
$ oc scale --replicas=<twice_the_number_of_replicas> \// <1>
  machineset <machine_set_name> \
  -n openshift-machine-api
----

> <1> For example, if the original number of replicas in the compute machine set was `2`, scale the replicas to `4`.

Reviewer: ```suggestion
<1> For example, if the original number of replicas in the compute machine set is `2`, scale the replicas to `4`.
```

Reviewer: Do we need to explain why we need to scale to double, or would a typical user know?

Author: No I do not believe it is necessary to explain why. If we get feedback after this docs go live that there is confusion, then we can update accordingly.

. List the machines that are managed by the updated compute machine set by running the following command:
+
[source,terminal]
----
$ oc get -n openshift-machine-api machines -l machine.openshift.io/cluster-api-machineset=<machine_set_name>
----
+
When the new machines are in the `Running` phase, you can scale the compute machine set to the original number of replicas.
. Scale the compute machine set to the original number of replicas by running the following command:
+
[source,terminal]
----
$ oc scale --replicas=<original_number_of_replicas> \// <1>
  machineset <machine_set_name> \
  -n openshift-machine-api
----

> <1> For example, if the original number of replicas in the compute machine set was `2`, scale the replicas to `2`.

Reviewer: ```suggestion
<1> For example, if the original number of replicas in the compute machine set is `2`, scale the replicas to `2`.
```

Author: Done. Thank you.

----
$ oc get machinesets -n openshift-machine-api
----
. Edit the first compute machine set by running the following command:
+
[source,terminal]
----
$ oc edit machineset <machineset_name> -n openshift-machine-api
----
. Note the value of `spec.replicas`, as you need it when scaling up the machine set to apply the changes.
. Configure the compute machine set to use the the first failure domain by adding the following to the `spec.template.spec.providerSpec.value` stanza:
+
.Example compute machine set with Nutanix failure domains
[source,yaml]
----
apiVersion: machine.openshift.io/v1
kind: MachineSet
metadata:
  creationTimestamp: null
  labels:
    machine.openshift.io/cluster-api-cluster: <cluster_name>
  name: <machineset_name>
  namespace: openshift-machine-api
spec:
  replicas: 2
# ...
  template:
    spec:
# ...
      providerSpec:
        value:
          apiVersion: machine.openshift.io/v1
          failureDomain:
            name: <failure_domain_name_1>
          cluster:
            type: uuid
            uuid: <prism_element_uuid_1>
          subnets:
          - type: uuid
            uuid: <prism_element_subnet_uuid_1>
# ...
----
. Save your changes.

> . Scale down the compute machine set to `0` by running the following command:

Reviewer 2: That's what we have in [docs](https://docs.openshift.com/container-platform/4.14/machine_management/modifying-machineset.html) for any change to machine sets, currently. Quite possible it's outdated; is there a better method to recommend or is it somehow handled automagically now?

> . Scale down the compute machine set to `0` by running the following command:

Reviewer 2: Closing this out: Joel suggested a better way to do this. I will update the content at [Modifying a compute machine set](https://docs.openshift.com/container-platform/4.14/machine_management/modifying-machineset.html#machineset-modifying_modifying-machineset), and then we can use those updates for this procedure. This can happen in parallel and shouldn't hold up this PR.

// Module included in the following assemblies:
//
// * post_installation_configuration/adding-nutanix-failure-domains.adoc

:_mod-docs-content-type: PROCEDURE
[id="post-installation-adding-nutanix-failure-domains-compute-machines_{context}"]
= Distributing compute machines across failure domains

You can distribute compute machines across Nutanix failure domains by:

* Modifying existing compute machine sets.
* Creating new compute machine sets.

The following procedure details how to distribute compute machines across failure domains by modify existing control plane machines sets. For more information on creating a control plane machine set, see xref:../machine_management/creating_machinesets/creating-machineset-nutanix.adoc#creating-machineset-nutanix[Creating a compute machine set on Nutanix].

.Prerequisites

> * You have added the failure domains to the cluster's Infrastructure custom resource definition (CRD).

Reviewer 3: Should change to:

You have added the failure domains configuration to the cluster's Infrastructure custom resource (CR).

Author: Done. Thank you.

.Procedure

> . Run the following command to view the cluster's Infrastructure CRD.

Reviewer 3: Change "CRD" to "CR".

Author: Made this change across the board. Thank you.

+
[source,terminal]
----
$ oc describe infrastructures.config.openshift.io cluster
----
. For each failure domain (`platformSpec.nutanix.failureDomains`), note the cluster's UUID, name, and subnet object UUID. These values are required to add a failure domain to a compute machine set.
. List the compute machine sets in your cluster by running the following command:
+
[source,terminal]
----
$ oc get machinesets -n openshift-machine-api
----
. Edit the first compute machine set by running the following command:
+
[source,terminal]
----
$ oc edit machineset <machineset_name> -n openshift-machine-api
----
. Note the value of `spec.replicas`, as you need it when scaling up the machine set to apply the changes.
. Configure the compute machine set to use the the first failure domain by adding the following to the `spec.template.spec.providerSpec.value` stanza:
+
[NOTE]
====

> Be sure that the values you specify for the `cluster` and `subnet` fields match the values that were configured in the `failureDomains` stanza when those that were configured in the cluster's Infrastructure custom resource definition (CRD).

Reviewer 3: Change "CRD" to "CR".
